Manually setting module to `__main__` for callables causes errors starting in 0.3.5

Context: Our code uses dill to serialize callables only under certain circumstances, including when the callable's module is `__main__`. We want to be able to test this at an integration level, using dill on a function where `func.__module__ == __main__`.

Prior to the release of 0.3.5, we did this by manually setting the `__module__` attribute. However, our tests have been failing since the release of 0.3.5.

I'm not sure if this is supposed to be an allowed use case. But is there supposed to be another way that I make it look like my functions were defined in `__main__` (for the part our code manages) such that dill can also handle it, when really I'm running under pytest or another testing framework?

MCVE: In an environment with pytest and dill installed:

```python
# test_dill.py

import dill

def setup_module():
    print(f"\nDill version: {dill.__version__}")


def external_func_pretending_to_be_main():
    return


def test_dill_lambda():
    print()  # cleaner output
    func = lambda x: x
    func.__module__ = "__main__"
    dill.dumps(func)


def test_dill_func():
    print() # cleaner output
    func = external_func_pretending_to_be_main
    func.__module__ = "__main__"
    dill.dumps(func)
```

```bash
pytest -vs test_dill.py
```

Output in 0.3.4:

```
test_dill.py::test_dill_lambda
Dill version: 0.3.4

PASSED
test_dill.py::test_dill_func
PASSED
```

Output in 0.3.5.1

```
test_dill.py::test_dill_lambda
Dill version: 0.3.5.1

FAILED
test_dill.py::test_dill_func
FAILED
```

Tracebacks, as reported by pytest:

<details><summary>test_dill_lambda</summary>

```pytb
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:364: in dumps
    dump(obj, file, protocol, byref, fmode, recurse, **kwds)#, strictio)
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:336: in dump
    Pickler(file, protocol, **_kwds).dump(obj)
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:620: in dump
    StockPickler.dump(self, obj)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:487: in dump
    self.save(obj)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:1963: in save_function
    _save_with_postproc(pickler, (_create_function, (
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:1154: in _save_with_postproc
    pickler._batch_setitems(iter(source.items()))
../../mambaforge/envs/dill/lib/python3.10/pickle.py:998: in _batch_setitems
    save(v)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:603: in save
    self.save_reduce(obj=obj, *rv)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:717: in save_reduce
    save(state)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:1251: in save_module_dict
    StockPickler.save_dict(pickler, obj)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:972: in save_dict
    self._batch_setitems(obj.items())
../../mambaforge/envs/dill/lib/python3.10/pickle.py:998: in _batch_setitems
    save(v)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:603: in save
    self.save_reduce(obj=obj, *rv)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:717: in save_reduce
    save(state)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:1251: in save_module_dict
    StockPickler.save_dict(pickler, obj)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:972: in save_dict
    self._batch_setitems(obj.items())
../../mambaforge/envs/dill/lib/python3.10/pickle.py:998: in _batch_setitems
    save(v)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:603: in save
    self.save_reduce(obj=obj, *rv)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:687: in save_reduce
    save(cls)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:1838: in save_type
    _save_with_postproc(pickler, (_create_type, (
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:1140: in _save_with_postproc
    pickler.save_reduce(*reduction, obj=obj)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:692: in save_reduce
    save(args)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
../../mambaforge/envs/dill/lib/python3.10/pickle.py:902: in save_tuple
    save(element)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:1251: in save_module_dict
    StockPickler.save_dict(pickler, obj)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:972: in save_dict
    self._batch_setitems(obj.items())
../../mambaforge/envs/dill/lib/python3.10/pickle.py:998: in _batch_setitems
    save(v)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:1963: in save_function
    _save_with_postproc(pickler, (_create_function, (
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:1154: in _save_with_postproc
    pickler._batch_setitems(iter(source.items()))
../../mambaforge/envs/dill/lib/python3.10/pickle.py:998: in _batch_setitems
    save(v)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:2004: in save_function
    StockPickler.save_global(pickler, obj, name=name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <dill._dill.Pickler object at 0x1082f2e90>
obj = <function make_set_closure_cell.<locals>.set_closure_cell at 0x1053c6290>
name = 'make_set_closure_cell.<locals>.set_closure_cell'

    def save_global(self, obj, name=None):
        write = self.write
        memo = self.memo

        if name is None:
            name = getattr(obj, '__qualname__', None)
        if name is None:
            name = obj.__name__

        module_name = whichmodule(obj, name)
        try:
            __import__(module_name, level=0)
            module = sys.modules[module_name]
            obj2, parent = _getattribute(module, name)
        except (ImportError, KeyError, AttributeError):
>           raise PicklingError(
                "Can't pickle %r: it's not found as %s.%s" %
                (obj, module_name, name)) from None
E           _pickle.PicklingError: Can't pickle <function make_set_closure_cell.<locals>.set_closure_cell at 0x1053c6290>: it's not found as attr._compat.make_set_closure_cell.<locals>.set_closure_cell

../../mambaforge/envs/dill/lib/python3.10/pickle.py:1071: PicklingError
```

</details>

<details><summary>test_dill_func</summary>

```pytb
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:364: in dumps
    dump(obj, file, protocol, byref, fmode, recurse, **kwds)#, strictio)
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:336: in dump
    Pickler(file, protocol, **_kwds).dump(obj)
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:620: in dump
    StockPickler.dump(self, obj)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:487: in dump
    self.save(obj)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:1963: in save_function
    _save_with_postproc(pickler, (_create_function, (
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:1154: in _save_with_postproc
    pickler._batch_setitems(iter(source.items()))
../../mambaforge/envs/dill/lib/python3.10/pickle.py:998: in _batch_setitems
    save(v)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:603: in save
    self.save_reduce(obj=obj, *rv)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:717: in save_reduce
    save(state)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:1251: in save_module_dict
    StockPickler.save_dict(pickler, obj)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:972: in save_dict
    self._batch_setitems(obj.items())
../../mambaforge/envs/dill/lib/python3.10/pickle.py:998: in _batch_setitems
    save(v)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:603: in save
    self.save_reduce(obj=obj, *rv)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:717: in save_reduce
    save(state)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:1251: in save_module_dict
    StockPickler.save_dict(pickler, obj)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:972: in save_dict
    self._batch_setitems(obj.items())
../../mambaforge/envs/dill/lib/python3.10/pickle.py:998: in _batch_setitems
    save(v)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:603: in save
    self.save_reduce(obj=obj, *rv)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:687: in save_reduce
    save(cls)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:1838: in save_type
    _save_with_postproc(pickler, (_create_type, (
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:1140: in _save_with_postproc
    pickler.save_reduce(*reduction, obj=obj)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:692: in save_reduce
    save(args)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
../../mambaforge/envs/dill/lib/python3.10/pickle.py:902: in save_tuple
    save(element)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:1251: in save_module_dict
    StockPickler.save_dict(pickler, obj)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:972: in save_dict
    self._batch_setitems(obj.items())
../../mambaforge/envs/dill/lib/python3.10/pickle.py:998: in _batch_setitems
    save(v)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:1963: in save_function
    _save_with_postproc(pickler, (_create_function, (
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:1154: in _save_with_postproc
    pickler._batch_setitems(iter(source.items()))
../../mambaforge/envs/dill/lib/python3.10/pickle.py:998: in _batch_setitems
    save(v)
../../mambaforge/envs/dill/lib/python3.10/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
../../mambaforge/envs/dill/lib/python3.10/site-packages/dill/_dill.py:2004: in save_function
    StockPickler.save_global(pickler, obj, name=name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <dill._dill.Pickler object at 0x10849f910>
obj = <function make_set_closure_cell.<locals>.set_closure_cell at 0x1053c6290>
name = 'make_set_closure_cell.<locals>.set_closure_cell'

    def save_global(self, obj, name=None):
        write = self.write
        memo = self.memo

        if name is None:
            name = getattr(obj, '__qualname__', None)
        if name is None:
            name = obj.__name__

        module_name = whichmodule(obj, name)
        try:
            __import__(module_name, level=0)
            module = sys.modules[module_name]
            obj2, parent = _getattribute(module, name)
        except (ImportError, KeyError, AttributeError):
>           raise PicklingError(
                "Can't pickle %r: it's not found as %s.%s" %
                (obj, module_name, name)) from None
E           _pickle.PicklingError: Can't pickle <function make_set_closure_cell.<locals>.set_closure_cell at 0x1053c6290>: it's not found as attr._compat.make_set_closure_cell.<locals>.set_closure_cell

../../mambaforge/envs/dill/lib/python3.10/pickle.py:1071: PicklingError
```

Thanks for working on this @anivegesana (and sorry for slow replies -- the affected project is passion project; day job gets in the way).

Using `recurse=True` seems to be a workaround (on master or on #486). Without `recurse=True`, #486 (at c81dfe54c7462dde969f397bccaa19b71b832f82) leads to a different error in the same spot, both in my actual tests in the MCVE above. I'm getting:

```pytb
TypeError: cannot pickle 'EncodedFile' object
```

<details> <summary>Full error (from pytest, just showing <code>test_dill_func</code> from above)</summary>

```pytb
____________________________________ test_dill_func ____________________________________

    def test_dill_func():
        print() # cleaner output
        func = external_func_pretending_to_be_main
        func.__module__ = "__main__"
>       dill.dumps(func)

test_dill.py:24:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
dill/dill/_dill.py:364: in dumps
    dump(obj, file, protocol, byref, fmode, recurse, **kwds)#, strictio)
dill/dill/_dill.py:336: in dump
    Pickler(file, protocol, **_kwds).dump(obj)
dill/dill/_dill.py:620: in dump
    StockPickler.dump(self, obj)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:487: in dump
    self.save(obj)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
dill/dill/_dill.py:2024: in save_function
    _save_with_postproc(pickler, (_create_function, (
dill/dill/_dill.py:1206: in _save_with_postproc
    pickler._batch_setitems(iter(source.items()))
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:997: in _batch_setitems
    save(v)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:603: in save
    self.save_reduce(obj=obj, *rv)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:717: in save_reduce
    save(state)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
dill/dill/_dill.py:1312: in save_module_dict
    StockPickler.save_dict(pickler, obj)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:971: in save_dict
    self._batch_setitems(obj.items())
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:997: in _batch_setitems
    save(v)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:603: in save
    self.save_reduce(obj=obj, *rv)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:717: in save_reduce
    save(state)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
dill/dill/_dill.py:1312: in save_module_dict
    StockPickler.save_dict(pickler, obj)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:971: in save_dict
    self._batch_setitems(obj.items())
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:997: in _batch_setitems
    save(v)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:603: in save
    self.save_reduce(obj=obj, *rv)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:717: in save_reduce
    save(state)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
dill/dill/_dill.py:1312: in save_module_dict
    StockPickler.save_dict(pickler, obj)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:971: in save_dict
    self._batch_setitems(obj.items())
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:997: in _batch_setitems
    save(v)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
dill/dill/_dill.py:1312: in save_module_dict
    StockPickler.save_dict(pickler, obj)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:971: in save_dict
    self._batch_setitems(obj.items())
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:997: in _batch_setitems
    save(v)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:603: in save
    self.save_reduce(obj=obj, *rv)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:717: in save_reduce
    save(state)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
dill/dill/_dill.py:1312: in save_module_dict
    StockPickler.save_dict(pickler, obj)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:971: in save_dict
    self._batch_setitems(obj.items())
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:997: in _batch_setitems
    save(v)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:603: in save
    self.save_reduce(obj=obj, *rv)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:717: in save_reduce
    save(state)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
dill/dill/_dill.py:1312: in save_module_dict
    StockPickler.save_dict(pickler, obj)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:971: in save_dict
    self._batch_setitems(obj.items())
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:997: in _batch_setitems
    save(v)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:603: in save
    self.save_reduce(obj=obj, *rv)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:717: in save_reduce
    save(state)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:560: in save
    f(self, obj)  # Call unbound method with explicit self
dill/dill/_dill.py:1312: in save_module_dict
    StockPickler.save_dict(pickler, obj)
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:971: in save_dict
    self._batch_setitems(obj.items())
../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:997: in _batch_setitems
    save(v)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <dill._dill.Pickler object at 0x10e67f820>
obj = <_io.TextIOWrapper name="<_io.FileIO name=7 mode='rb+' closefd=True>" mode='r+' encoding='utf-8'>
save_persistent_id = True

    def save(self, obj, save_persistent_id=True):
        self.framer.commit_frame()

        # Check for persistent id (defined by a subclass)
        pid = self.persistent_id(obj)
        if pid is not None and save_persistent_id:
            self.save_pers(pid)
            return

        # Check the memo
        x = self.memo.get(id(obj))
        if x is not None:
            self.write(self.get(x[0]))
            return

        rv = NotImplemented
        reduce = getattr(self, "reducer_override", None)
        if reduce is not None:
            rv = reduce(obj)

        if rv is NotImplemented:
            # Check the type dispatch table
            t = type(obj)
            f = self.dispatch.get(t)
            if f is not None:
                f(self, obj)  # Call unbound method with explicit self
                return

            # Check private dispatch table if any, or else
            # copyreg.dispatch_table
            reduce = getattr(self, 'dispatch_table', dispatch_table).get(t)
            if reduce is not None:
                rv = reduce(obj)
            else:
                # Check for a class with a custom metaclass; treat as regular
                # class
                if issubclass(t, type):
                    self.save_global(obj)
                    return

                # Check for a __reduce_ex__ method, fall back to __reduce__
                reduce = getattr(obj, "__reduce_ex__", None)
                if reduce is not None:
>                   rv = reduce(self.proto)
E                   TypeError: cannot pickle 'EncodedFile' object

../../mambaforge/envs/dill-check/lib/python3.9/pickle.py:578: TypeError
```

--- a/dill/_dill.py
+++ b/dill/_dill.py
@@ -1130,13 +1130,35 @@ def _import_module(import_name, safe=False):
             return None
         raise
 
-def _locate_function(obj, pickler=None):
-    if obj.__module__ in ['__main__', None] or \
-            pickler and is_dill(pickler, child=False) and pickler._session and obj.__module__ == pickler._main.__name__:
-        return False
+# https://github.com/python/cpython/blob/a8912a0f8d9eba6d502c37d522221f9933e976db/Lib/pickle.py#L322-L333
+def _getattribute(obj, name):
+    for subpath in name.split('.'):
+        if subpath == '<locals>':
+            raise AttributeError("Can't get local attribute {!r} on {!r}"
+                                 .format(name, obj))
+        try:
+            parent = obj
+            obj = getattr(obj, subpath)
+        except AttributeError:
+            raise AttributeError("Can't get attribute {!r} on {!r}"
+                                 .format(name, obj))
+    return obj, parent
 
-    found = _import_module(obj.__module__ + '.' + obj.__name__, safe=True)
-    return found is obj
+def _locate_function(obj, pickler=None):
+    module_name = getattr(obj, '__module__', None)
+    if module_name in ['__main__', None] or \
+            pickler and is_dill(pickler, child=False) and pickler._session and module_name == pickler._main.__name__:
+        return False
+    if hasattr(obj, '__qualname__'):
+        module = _import_module(module_name, safe=True)
+        try:
+            found, _ = _getattribute(module, obj.__qualname__)
+            return found is obj
+        except:
+            return False
+    else:
+        found = _import_module(module_name + '.' + obj.__name__, safe=True)
+        return found is obj
 
 
 def _setitems(dest, source):

